File: frontend/src/service/userService.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { SaveOrderDto } from './order';

export interface UserDto {
  id: number;
  name: string;
  email: string;
}
export interface SaveUserDto {
  name: string;
  email: string;
  password: string;
}
export interface Page<T> {
  content: T[];
  totalPages: number;
  totalElements: number;
  size: number;
  number: number;
}
export interface LoginDto {
  email: string;
  password: string;
}

export interface AuthResponse {
  token: string;
}

@Injectable({
  providedIn: 'root',
})
export class UserService {
  private apiUrl = 'http://localhost:8080/users';

  constructor(private http: HttpClient) {}

  getAll(page: number = 0, size: number = 4): Observable<Page<UserDto>> {
    return this.http.get<Page<UserDto>>(`${this.apiUrl}?page=${page}&size=${size}`);
  }

  getById(id: number): Observable<UserDto> {
    return this.http.get<UserDto>(`${this.apiUrl}/${id}`);
  }

  register(user:SaveUserDto): Observable<UserDto> {
    return this.http.post<UserDto>(`${this.apiUrl}/register`, user);
  }

  login(credentials: LoginDto): Observable<AuthResponse> {
    return this.http.post<AuthResponse>(`${this.apiUrl}/login`, credentials);
  }

}

File: backend/User-Service/src/main/java/org/example/userservice/service/UserService.java
package org.example.userservice.service;

import org.example.userservice.dto.AuthResponseDto;
import org.example.userservice.dto.LoginDto;
import org.example.userservice.dto.SaveUserDto;
import org.example.userservice.dto.UserDto;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

public interface UserService {

    UserDto register(SaveUserDto dto);

    AuthResponseDto login(LoginDto dto);

    UserDto getById(Long id);

    Page<UserDto> getAll(Pageable pageable);
}
